### ISSUE TITLE
Improvement in theming: adapt colour to app

### ISSUE THREAD

**Issue by GithbUser** [florisboard/florisboard#226]
When adapt colour to app opt-in is turned on
There is  lag sometimes when opening an app like signal or WhatsApp
But this doesn't happen when the option is turned off
Also sometimes it switches to other keyboards
 
![Screenshot_20210123-102218284](https://user-images.githubusercontent.com/75420364/105569276-b0587300-5cfd-11eb-8a68-c9472de767cd.jpg)

There is also some problems adapting keyboard in vlc app while using adapt colour to app option

![Screenshot_20210123-102205791](https://user-images.githubusercontent.com/75420364/105569309-f44b7800-5cfd-11eb-8d59-e5a32e19cdd6.jpg)

I hope my suggestion is understandable for you 
Thank you very much for developing Florisboard
Version-0.3.4
Android 10
 resurrection Remix os vanilla

**Comment by Glitchy-Tozier** [comment]
1. I haven't noticed the Lag (however, I've only used the adapting themes for half an hour now.)
2. I _can_ however reproduce the transparency-bug.
![Imagepipe_9.jpg](https://user-images.githubusercontent.com/59611881/105570792-ac534200-5d4b-11eb-85ae-344ab56a0e13.jpg)


I also have a little improvement idea, let me just post this here:
Sometimes, key-colour and backgroud-color are the same. This is what happened to me in Signal:
![Screenshot_20210123-064707_Signal.jpg](https://user-images.githubusercontent.com/59611881/105570840-ea506600-5d4b-11eb-8652-280579590cb1.jpg)

**Comment by GithbUser** [comment]
@Glitchy-Tozier 
Yes you are right about same key and background color in some apps like dialer ,messenging
 maybe the lags i m facing because i dont use gapps or microg 
But a few times the lag also switches to other keyboard like aosp keyboard 
![Screenshot_20210123-134730935](https://user-images.githubusercontent.com/75420364/105573242-265dd980-5d00-11eb-9d2d-3d6c6937caba.jpg)
![Screenshot_20210123-134711632](https://user-images.githubusercontent.com/75420364/105573244-2b228d80-5d00-11eb-8e9b-b349ae8696b7.jpg)

**Comment by Glitchy-Tozier** [comment]
I can see the issue with the top screenshot.
But could you describe what's wfong woth the bottom (green) screenshot?

(sgreenshot)

**Comment by GithbUser** [comment]
@Glitchy-Tozier 
I am not saying its an issue
But requesting for improvement in theme

**Comment by Glitchy-Tozier** [comment]
I see. I think the more precise we are, the easier it is for @patrickgold to implement stuff.

So here's a few other small things, that could be improved:
1. **I have the feeling that the color-extraction doesn't fully work yet.**
On F-Droid, KISS launcher, Firefox Klar (Focus), and Fennec F-Droid (Firefox), FlorisBoard stays the default color (white, grey, green)
However, AnySoftKeyboard (ASK) does indeed change color for all those apps.
2. **Key pop-ups should change color as well.**
I actually enjoy keeping the bright pop-ups, bht it makes sense for their color to change as well.
3. **The enter Key turns gree , regardless of Keyboard-color.**
That is acceptable imo, if you are palnning on meeping FlorisBoard's green theming somewhat omnipresent. Your choice.
4. **Sometimes, when switching applications...**
  - the previous' keyboard color either quickly flickers before being replaced.
  - Also, it sometimes just stays the wrong color.
  - Also, sometimes switching between apps messes up the new colors. Only after closing&re-opening the keyboard corrects itself.

**Comment by patrickgold** [comment]
Thanks for your improvement suggestions, I really appreciate it!

Adaptive theme colors is a feature which I have to say is implemented but works very inconsistently. This is due to the fact that there's such a variety of different ways app devs implement their themes and I only test for the default "Android way", because this is the most popular among them.

The reason why the colors are sometimes ok and sometimes messed up is because of the way I get the colors:

At every `onStartInputView()` event I receive I get the package name of the tagret app. With this, I am asking the Android system about the launch intent for this app. This is normally used by launchers to get the app name and icon, but I can also extract the `theme` attribute from there. Then I check this theme if it has certain standard color attributes, in specific `colorPrimary` and `colorPrimaryDark`. If these colors do not exist, FlorisBoard will default to its own theme colors (what you've configured in the Theme settings).

Now the transparency thing is not directly a bug, because it means the target app has specified a transparent color in this field (even though there should normally not be a transparent color there). What I could do against this is forcing the alpha channel to 100% transparency, so this does not happen at least.

Then there's the thing with the background color being the same as the key color. There's absolutely nothing I can do against this, as this means the target app has specified both `colorPrimary` and `colorPrimaryDark` to the same color. I cannot identify if this is an error or on purpose, so I will just apply this.

AnySoftKeyboard has a slightly better approach, which prevents the transparency bug and sometimes gives more accurate colors. Still, far from perfect (at least on a system dark themed device) and is often times messed up like my implementation, especially though in Google Apps.

Now to the popups: I have to agree I kinda forgot about the popups and that they also should get affected by the adaptive themeing, will add this to the todo list.

Ok I got distracted and @Glitchy-Tozier wrote a comment in the meantime, so I will address the points I haven't talked about yet:

> The enter Key turns green, regardless of Keyboard-color.

The enter key does only turn green if the adaptive theme fails to find colors and if you haven't chnaged the enter key color in the Theme settings.

> Sometimes, when switching applications...

Hm strange, this is definitely a bug I have to investigate in. COuld be because the launch intent request crashes somewhere, but I am unsure.

> The colors of the android-buttons at the bottom of the screen sometimes really clash with the keyboard's colors. Example:

This is because it is bound to the current theme value rather than the adaptive one, will add this to the todo list.

**Comment by Glitchy-Tozier** [comment]
> What I could do against this is forcing the alpha channel to 100% transparency, so this does not happen at least.

Sounds good to me.

> Then there's the thing with the background color being the same as the key color.

Right, I checked, ASK has the same problem.
Is there a way of saying "If the fore- & background-colors are too similar to each other, tweak one of them?

> The enter key does only turn green if the adaptive theme fails to find colors and if you haven't changed the enter key color in the Theme settings.
I didn't change anything about the themes except toggling "Adapt Colors to app".

I think, I might have not beet very clear about what I meant. I meant "the enter-key turns green **while being pressed**".

> This is because it is bound to the current theme value rather than the adaptive one, will add this to the todo list. 

👍

**Comment by patrickgold** [comment]
>  Is there a way of saying "If the fore- & background-colors are too similar to each other, tweak one of them?

Definitely if I look at the material Guidelines of Google, which describe exactly how the `colorPrimary` and `colorPrimaryDark` should be apart. So if I at least get one of these, I can somehow "calculate" the other.

> I think, I might have not beet very clear about what I meant. I meant "the enter-key turns green while being pressed".

Oh this is something I forgot to tweak it seems to me. Normally this should also adapt to the target app color.

Another idea I have but currently don't know if it is too much work is to make a predefined list for the most popular apps (on GPlay as well as in the FOSS landscape), so at least these apps have a proper "adaptive theme". I would create a list for light and dark theme, as it seems there is quite a difference often. Though i don't know if this effort is really worth the outcome, what do you think about this?

**Comment by Glitchy-Tozier** [comment]
> Though i don't know if this effort is really worth the outcome, what do you think about this?

Honestly, if I was the developer, I wouldn't do it.

Two reasons:
1. It may not be worth it, and there are more important projects to focus on.
2. I think people _know_ that variable themes are going to be a little janky.
Personally, I think it's kind of fun to go into a new app and think "I wonder what that does to the keyboard"

I think people don't expect app-dependant themes to be amazing. Of course they want them to be decent, but if you are able to fix the above issues (especially the monotone keyboard and the transparency-issue), that is good enough.  :)


Also, If you ever decide on doing the app list, I wouldn't suggest doing it yourself.
I would be lazy and kind of croud-source it. Let users submit designs.
That's just a future thought though, after speech-recognition has been implemented, after hundrets of top-bars have been added, and after word prediction is good enough for you to not need letter-keys anymore (except to start the first word)   ;)

**Comment by patrickgold** [comment]
Hmm you are right there are more important projects than this. I will implement the improvements this issue brought up and will call it a day.

> Also, If you ever decide on doing the app list, I wouldn't suggest doing it yourself.
I would be lazy and kind of croud-source it. Let users submit designs.

Of course, I don't have the time and desire to install thousands of apps just to compile this list, I would have definitely crowd-sourced it.

>  That's just a future thought though, after speech-recognition has been implemented, after hundrets of top-bars have been added, and after word prediction is good enough for you to not need letter-keys anymore (except to start the first word) ;)

:)

**Comment by patrickgold** [comment]
Above PR improves adaptive themeing. For details see the text in the PR itself.

I didn't experience any lags/crashes/return to defaults on my emulator device as well as on my physical devices, so I see this as fixed. Although if there should still be severe lags for you in v0.,3.5, feel free to reopen this issue! Also, I was unable to create a good color calculation for same keyboard&key background colors, so for now this is still in. The overhaul has fixed and improved the colors for a lot of apps though, some apps like WhatsApp, Discord, Signal (at least dark theme), etc. look really really good now. Google apps are acceptable (at least most of them), Fiexfox works well, VLC now has a black instead of transparent background, etc.

**Comment by Glitchy-Tozier** [comment]
@patrickgold Nice job with the themes (and the update in general)!  :)
A few random things I would like to tell or ask you:


1. I finally was able to make a screenshot of the color-carry-over-bug
Original screen:
![Screenshot_20210126-162027_KISS launcher.jpg](https://user-images.githubusercontent.com/59611881/105864746-8b604a80-5ff2-11eb-9cc6-d7187acce1aa.jpg)
Bug:
![Screenshot_20210126-160233_Firefox Klar.jpg](https://user-images.githubusercontent.com/59611881/105864779-931fef00-5ff2-11eb-8ce2-8450bada7730.jpg)
Keyboard after closing&re-opening it:
![Screenshot_20210126-160239_Firefox Klar.jpg](https://user-images.githubusercontent.com/59611881/105864881-aa5edc80-5ff2-11eb-9efc-aeca93fcf418.jpg)

2. Somehow, you made the themes work in way more apps! For example, Fennec now works. And Whatsapp looks way better! (Signal's still ugly though)

3. Is this an inteded thing?
In Github, the keyboard is completely white, while pop-ups are blue. It looms nice, but why and how is the secondary color not visible in the keyboard?
![Screenshot_20210126-161456_GitHub.jpg](https://user-images.githubusercontent.com/59611881/105865460-3f61d580-5ff3-11eb-915a-a25a97f9a44b.jpg)

4. What do you think about letting the user set the color-preferences for certain apps themselves?
(Or at least disable adaptive theming cor certain apps, for now)
What made me think of this is that in some cases the user _really_ needs to change the keyboard colors.

Example:
Look at my launcher. See how nice it is.  :)
![Screenshot_20210126-161946_KISS launcher.jpg](https://user-images.githubusercontent.com/59611881/105866016-ce6eed80-5ff3-11eb-8847-cc205562ce46.jpg)

Now the keyboard pops up.
![Screenshot_20210126-162027_KISS launcher.jpg](https://user-images.githubusercontent.com/59611881/105866056-db8bdc80-5ff3-11eb-8bbd-6353d4ae1fad.jpg)

I think, some people may have similar problems.

**Comment by patrickgold** [comment]
> Nice job with the themes (and the update in general)! :)

@Glitchy-Tozier Glad you like the update :)

> 1. I finally was able to make a screenshot of the color-carry-over-bug

Strange. I am explicitly resetting the adaptive theme every time the package name changes. As you can see, the key background color changes but not the keyboard background color. I suspect that the InputView gets the new color but fails to redraw its background somehow. I will investigate into this.

> 2. Somehow, you made the themes work in way more apps! For example, Fennec now works. And Whatsapp looks way better! (Signal's still ugly though)

This is because I am looking at the Application too in case the launch intent does not give back a theme. seems that quite a few apps seem to not have a theme in their startup activity but in the application tag itself.

> 3. Is this an inteded thing?
In Github, the keyboard is completely white, while pop-ups are blue. It looms nice, but why and how is the secondary color not visible in the keyboard?

I guess this is what GitHub did in their Android app:

```
colorPrimary = white
colorPrimaryDark = white
colorAccent = blue
```

If looking at how their app and navigation bar / status bar are colored, I guess my above assumption is right. The problem is if an app has defined the colors this way, they will also be this way in FlorisBoard. This is what colors do what to FlorisBoard:

```
colorPrimary: key background
colorPrimaryDark: keyboard background
colorAccent: popup background
```

The only thing that could potentially fix the issue in the GitHub app is by calculating broken colors (for instance making `colorPrimaryDark` more dark), but like I said I failed to deliver accurate and good results here, maybe I will try a different approach and get this up and running.

> 4. What do you think about letting the user set the color-preferences for certain apps themselves?
(Or at least disable adaptive theming cor certain apps, for now)
What made me think of this is that in some cases the user really needs to change the keyboard colors.

Sounds like a nice idea but requires a lot of work (UI wise in the Settings) to get this up and running for a workaround which should not be needed if adaptive theme works well for all apps.

**Comment by Glitchy-Tozier** [comment]
> I will investigate into this.

Do some detective-work!


> ... but like I said I failed to deliver accurate and good results here, maybe I will try a different approach and get this up and running.

What approach did you try?
Is it possible to calculate some measure of color-difference with whatever values you receive from the app?

> Sounds like a nice idea but requires a lot of work (UI wise in the Settings) to get this up and running for a workaround which should not be needed if adaptive theme works well for all apps.

In this case the app didn't do anything wrong. My launcher is a green app. (https://github.com/Neamar/KISS).
It's just that I changed all the colors. (And apparently the app doesn't change the color-values it returns to keyboards)

An easier option could be to just enable the user to turn off adaptive themes for applications.
That being said, it's not an extremely important addition, so if you want to do this in way later updates or don't want to add it at all, just tell me and I'll turn off adaptive themes for this launcher-skin.

**Comment by patrickgold** [comment]
> What approach did you try?
Is it possible to calculate some measure of color-difference with whatever values you receive from the app?

It is possible to calculate one color which is based on another color by subtracting / adding values for each color channel (red, green, blue), but you have to get these difference numbers right. Also detecting when to add and when to subtract is sometimes difficult and each channel must stay with in 0x00 and 0xFF to be a valid channel value.

> In this case the app didn't do anything wrong. My launcher is a green app. (https://github.com/Neamar/KISS).
It's just that I changed all the colors. (And apparently the app doesn't change the color-values it returns to keyboards)

What I said previously and again, I want to automatically avoid such cases by forcing one color to be lighter/darker so these same key/keyboard colors should not happen, regardless what the target app gives me.

> An easier option could be to just enable the user to turn off adaptive themes for applications.
That being said, it's not an extremely important addition, so if you want to do this in way later updates or don't want to add it at all, just tell me and I'll turn off adaptive themes for this launcher-skin.+

What I can do what is relatively easy to implement is adding a text box dialog to the theme preferences where you can enter all excluded package names, comma-separated. Of course not the most elegant solution but it is very easy to add for me (requires literally just a textbox and an if statement if the package exists within the preference string) and users who want to specifically disable adaptive themes for some apps can do so.

**Comment by Glitchy-Tozier** [comment]
> Like so for instance: fr.neamar.kiss,com.example.anotherapp,com.google.someapp. Leaving this empty does not exclude any packages.

I would be happy with that  :)


> What I said previously and again, I want to automatically avoid such cases by forcing one color to be lighter/darker so these same key/keyboard colors should not happen, regardless what the target app gives me.

I think there's a small misunderstanding.

What I'm saying is NOT "I don't like that fore- & background-color aren't different"
(even though this also is the case.)

What I wanted to say is: Even though I set the launcher to something purple-ish, the keyboard will always be green because that is the color that the launcher makes it be.

**Comment by patrickgold** [comment]
> What I wanted to say is: Even though I set the launcher to something purple-ish, the keyboard will always be green because that is the color that the launcher makes it be.

That is because Android system themes weren't really thought for runtime modification. As the KISS launcher seems to be a green colored launcher on default, this is also what the compiler baked into the Android system theme for the KISS launcher. You changing the colors within KISS launcher is kinda a dynamic override which circumvents the default theme resources. I cannot retrieve the dynamic resources, only the public baked-in theme of KISS. So it is always returning green. Within FlorisBoard I am also avoiding using the Android system theme model in the keyboard view to make dynamic themeing even possible.

**Comment by Glitchy-Tozier** [comment]
Exactly! And because of this, because KISS (and probably other launchers as well) will not send you the correct color, I think that exception-textbox is a good idea.

**Comment by Glitchy-Tozier** [comment]
> Strange. I am explicitly resetting the adaptive theme every time the package name changes. As you can see, the key background color changes but not the keyboard background color. I suspect that the InputView gets the new color but fails to redraw its background somehow. I will investigate into this.

Another screenshot: ![20210127_084444.jpg](https://user-images.githubusercontent.com/59611881/105959367-189bb180-607c-11eb-9ef7-55bed5f9ff4e.jpg)
This time it happened when going from Signal to (the green) launcher.

It seems like you _did_ fix some parts of the bug, as only ever the border is wrong. I didn't find any instances of the keys themselves having the wrong color.

**Comment by patrickgold** [comment]
> Does that even make sense? Does this setting ...
... not do anything if you're using adaptive themes?
(The "Dark Mode Only"-setting)

It does: when the adaptive theme fails for an app, you will always get the dark variant of FlorisBoard in the above setting. Also, I made a different toggle for day and night to enable people to use adaptive themeing at day but always use the dark theme at night, or vice versa. Also some components like scrollbar thumb color and the press effect of some buttons are dependent on if FlorisBoard is currently in day or night mode.

> This time it happened when going from Signal to (the green) launcher.
It seems like you did fix some parts of the bug, as only ever the border is wrong. I didn't find any instances of the keys themselves having the wrong color.

Like I said I have to force invalidate the texture of the InputView background as setting the background color seems to not do this on all devices.

**Comment by Glitchy-Tozier** [comment]
> It does: when the adaptive theme fails for an app, you will always get the dark variant of FlorisBoard in the above setting. Also, I made a different toggle for day and night to enable people to use adaptive themeing at day but always use the dark theme at night, or vice versa. Also some components like scrollbar thumb color and the press effect of some buttons are dependent on if FlorisBoard is currently in day or night mode.

Understood.


> Like I said I have to force invalidate the texture of the InputView background as setting the background color seems to not do this on all devices.

I wasn't sure how much you knew, so I just thought it might be helpful to tell you that it didn't happen with buttons yet.

**Comment by Glitchy-Tozier** [comment]
1. Just to make this clear: Only the real keyboard has this partial bug, if it occurs.
The loading-keyboard-preview will keep ALL colors from the previous app. When the bug occurs, that is.

2. Should we reopen this issue to make it clear that this still is in devellopment?

**Comment by patrickgold** [comment]
Yes, it would be best to open a new issue for this, if possible. I have an idea how to fix this, but I am currently fixing the landscreen popup bug.

**Comment by Glitchy-Tozier** [comment]
@patrickgold Not sure where to post this, so I'll add my comment to this issue.

We once talked about sometimes themes having inverted color-brightness, in the sense that dark colors are where bright colors should be and vice versa.
Additionally we had the problem of Apps' light and dark colors being too similar.
For both of those issues a way of converting RGB to (perceived) brightness would begreat, and as it turns out, formulas for this kind of thing exist: https://stackoverflow.com/questions/596216/formula-to-determine-perceived-brightness-of-rgb-color#596243

Of course i'm not expecting you to stop what you're doing and immediately work on this, I just thought I'd mention it. :)

**Comment by patrickgold** [comment]
@Glitchy-Tozier Do you mean adaptive themes or the default themes with the brightness formula? I currently cannot directly remember what topic you are referencing in your comment.

**Comment by Glitchy-Tozier** [comment]
Adaptive themes, like the Gmail dark theme:
![Screenshot_20211222-100331_Gmail.jpg](https://user-images.githubusercontent.com/59611881/147066213-b311f4ac-b22c-450c-b1f2-db947c11df51.jpg)

**Comment by patrickgold** [comment]
So you mean that I should calculate the background based on the key color instead of relying on the app to tell me a correct color? Would make sense IMO, because often times the primary and primary variant colors are set incorrectly or read out faulty.

**Comment by Glitchy-Tozier** [comment]
Exactly, kind of figuring out whether the app told you the darker color in the correct position, and if it didn't, switching the colors.

~Additionally, using this technique, if an app tells you the same color twice, you could slightly change one or both of the colors.~
Actually, not that I think about it, you might not need brightness to detect cases like this. Just compare the RGB-values:
![Screenshot_20211222-211313_KISS launcher.jpg](https://user-images.githubusercontent.com/59611881/147150433-352f2baf-c770-405a-9060-8a246fff41df.jpg)
